File: apps/agent/src/lib/firebase-auth.ts
import crypto from "node:crypto";
import type { NextFunction, Request, Response } from "express";

/**
 * Firebase ID-token verification with zero extra dependencies.
 *
 * Firebase ID tokens are RS256 JWTs signed by Google's secure-token service.
 * We verify them by hand: fetch Google's public x509 certs, check the RS256
 * signature against the cert named by the token's `kid`, then validate the
 * issuer / audience / expiry claims.
 *
 * Config-gated: if FIREBASE_PROJECT_ID isn't set, isFirebaseConfigured() is
 * false and the middleware is a no-op — the app runs open (demo behaviour).
 * When configured, mutating/agent requests must carry a valid Bearer token.
 */

const CERT_URL =
  "https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com";

function projectId(): string {
  return process.env["FIREBASE_PROJECT_ID"] || process.env["NEXT_PUBLIC_FIREBASE_PROJECT_ID"] || "";
}

export function isFirebaseConfigured(): boolean {
  return projectId().length > 0;
}

export interface FirebaseUser {
  uid: string;
  email?: string;
  emailVerified?: boolean;
  name?: string;
}

// ── cert cache ──
let certCache: { certs: Record<string, string>; expiresAt: number } | null = null;

async function getCerts(): Promise<Record<string, string>> {
  const now = Date.now();
  if (certCache && certCache.expiresAt > now) return certCache.certs;
  const res = await fetch(CERT_URL);
  if (!res.ok) throw new Error(`firebase: failed to fetch certs (${res.status})`);
  const certs = (await res.json()) as Record<string, string>;
  // honour max-age so we refresh roughly when Google rotates keys
  const cc = res.headers.get("cache-control") ?? "";
  const m = /max-age=(\d+)/.exec(cc);
  const maxAgeMs = m && m[1] ? Number(m[1]) * 1000 : 3600_000;
  certCache = { certs, expiresAt: now + maxAgeMs };
  return certs;
}

function b64urlToBuffer(s: string): Buffer {
  return Buffer.from(s.replace(/-/g, "+").replace(/_/g, "/"), "base64");
}

function b64urlToJson<T>(s: string): T {
  return JSON.parse(b64urlToBuffer(s).toString("utf8")) as T;
}

export async function verifyFirebaseToken(token: string): Promise<FirebaseUser> {
  const pid = projectId();
  if (!pid) throw new Error("firebase: not configured");

  const parts = token.split(".");
  if (parts.length !== 3) throw new Error("firebase: malformed token");
  const [headerB64, payloadB64, sigB64] = parts as [string, string, string];

  const header = b64urlToJson<{ alg?: string; kid?: string }>(headerB64);
  if (header.alg !== "RS256") throw new Error("firebase: unexpected alg");
  if (!header.kid) throw new Error("firebase: missing kid");

  const certs = await getCerts();
  const cert = certs[header.kid];
  if (!cert) throw new Error("firebase: no matching public cert");

  // Verify the RS256 signature against the x509 cert's public key.
  const publicKey = new crypto.X509Certificate(cert).publicKey;
  const ok = crypto.verify(
    "RSA-SHA256",
    Buffer.from(`${headerB64}.${payloadB64}`),
    publicKey,
    b64urlToBuffer(sigB64),
  );
  if (!ok) throw new Error("firebase: bad signature");

  const payload = b64urlToJson<{
    aud?: string;
    iss?: string;
    sub?: string;
    exp?: number;
    iat?: number;
    email?: string;
    email_verified?: boolean;
    name?: string;
  }>(payloadB64);

  const nowSec = Math.floor(Date.now() / 1000);
  if (payload.aud !== pid) throw new Error("firebase: wrong audience");
  if (payload.iss !== `https://securetoken.google.com/${pid}`) throw new Error("firebase: wrong issuer");
  if (!payload.sub) throw new Error("firebase: missing subject");
  if (typeof payload.exp !== "number" || payload.exp <= nowSec) throw new Error("firebase: token expired");
  if (typeof payload.iat === "number" && payload.iat > nowSec + 300) throw new Error("firebase: token issued in the future");

  return {
    uid: payload.sub,
    ...(payload.email ? { email: payload.email } : {}),
    ...(payload.email_verified !== undefined ? { emailVerified: payload.email_verified } : {}),
    ...(payload.name ? { name: payload.name } : {}),
  };
}

function bearer(req: Request): string | null {
  const h = req.headers["authorization"];
  if (typeof h !== "string") return null;
  const m = /^Bearer\s+(.+)$/i.exec(h);
  return m && m[1] ? m[1] : null;
}

/**
 * Express middleware. No-op when Firebase isn't configured. When it is,
 * read-only (GET/HEAD/OPTIONS) and /auth/* requests pass through, but every
 * mutating/agent request must carry a valid Firebase ID token.
 */
export async function requireAuth(req: Request, res: Response, next: NextFunction): Promise<void> {
  if (!isFirebaseConfigured()) return next();

  const method = req.method.toUpperCase();
  const isRead = method === "GET" || method === "HEAD" || method === "OPTIONS";
  const isAuthRoute = req.path.startsWith("/auth") || req.path === "/health" || req.path === "/ready";

  const token = bearer(req);
  // Always attach the user when a valid token is present (even on reads).
  if (token) {
    try {
      (req as Request & { user?: FirebaseUser }).user = await verifyFirebaseToken(token);
      return next();
    } catch (err) {
      // An explicitly-bad token is always rejected.
      res.status(401).json({ error: "unauthorized", detail: err instanceof Error ? err.message : String(err) });
      return;
    }
  }

  if (isRead || isAuthRoute) return next();

  res.status(401).json({ error: "unauthorized", detail: "missing Firebase ID token" });
}

File: apps/agent-py/app/lib/firebase_auth.py
"""Firebase ID-token verification — port of apps/agent/src/lib/firebase-auth.ts.

Config-gated: if FIREBASE_PROJECT_ID isn't set, the middleware is a no-op.
Verifies RS256 ID tokens against Google's public x509 certs with PyJWT.
"""
from __future__ import annotations

import os
import time

import httpx
import jwt
from cryptography.x509 import load_pem_x509_certificate
from fastapi import Request
from fastapi.responses import JSONResponse

_CERT_URL = "https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com"
_cert_cache: dict = {"certs": {}, "expires_at": 0.0}


def _project_id() -> str:
    return os.environ.get("FIREBASE_PROJECT_ID") or os.environ.get("NEXT_PUBLIC_FIREBASE_PROJECT_ID") or ""


def is_firebase_configured() -> bool:
    return len(_project_id()) > 0


async def _get_certs() -> dict:
    now = time.time()
    if _cert_cache["certs"] and _cert_cache["expires_at"] > now:
        return _cert_cache["certs"]
    async with httpx.AsyncClient(timeout=15) as c:
        r = await c.get(_CERT_URL)
        r.raise_for_status()
        certs = r.json()
        cc = r.headers.get("cache-control", "")
        import re
        m = re.search(r"max-age=(\d+)", cc)
        max_age = int(m.group(1)) if m else 3600
    _cert_cache.update({"certs": certs, "expires_at": now + max_age})
    return certs


async def verify_firebase_token(token: str) -> dict:
    pid = _project_id()
    if not pid:
        raise ValueError("firebase: not configured")
    header = jwt.get_unverified_header(token)
    kid = header.get("kid")
    if not kid:
        raise ValueError("firebase: missing kid")
    certs = await _get_certs()
    cert_pem = certs.get(kid)
    if not cert_pem:
        raise ValueError("firebase: no matching public cert")
    public_key = load_pem_x509_certificate(cert_pem.encode()).public_key()
    payload = jwt.decode(
        token, public_key, algorithms=["RS256"], audience=pid,
        issuer=f"https://securetoken.google.com/{pid}",
    )
    if not payload.get("sub"):
        raise ValueError("firebase: missing subject")
    return {"uid": payload["sub"], "email": payload.get("email"),
            "emailVerified": payload.get("email_verified"), "name": payload.get("name")}


def _bearer(request: Request) -> str | None:
    h = request.headers.get("authorization")
    if not h:
        return None
    parts = h.split(None, 1)
    return parts[1] if len(parts) == 2 and parts[0].lower() == "bearer" else None


async def firebase_middleware(request: Request, call_next):
    """No-op unless configured. Reads + /auth/* pass; mutations need a token."""
    if not is_firebase_configured():
        return await call_next(request)
    method = request.method.upper()
    is_read = method in ("GET", "HEAD", "OPTIONS")
    path = request.url.path
    is_auth_route = path.startswith("/auth") or path in ("/health", "/ready")

    token = _bearer(request)
    if token:
        try:
            request.state.user = await verify_firebase_token(token)
            return await call_next(request)
        except Exception as err:  # noqa: BLE001
            return JSONResponse(status_code=401, content={"error": "unauthorized", "detail": str(err)})
    if is_read or is_auth_route:
        return await call_next(request)
    return JSONResponse(status_code=401, content={"error": "unauthorized", "detail": "missing Firebase ID token"})
